File: src/main/java/minic/ui/MiniCVisualModelFactory.java
package minic.ui;

import minic.uiapi.UiGlobalDataDto;
import minic.uiapi.UiStageDataDto;

import java.util.ArrayList;
import java.util.List;

/**
 * 根据 UI API DTO 生成 Visual Pane 展示项。
 */
public final class MiniCVisualModelFactory {
    /**
     * 创建当前阶段可视化项。
     *
     * @param stageData 当前阶段数据
     * @param globalData 全局数据
     * @return 可视化项
     */
    public List<MiniCVisualItem> create(UiStageDataDto stageData, UiGlobalDataDto globalData) {
        if (stageData == null) {
            return List.of(new MiniCVisualItem("等待开始观测会话", true));
        }
        List<String> source = sourceFor(stageData.stage(), stageData, globalData);
        ArrayList<MiniCVisualItem> items = new ArrayList<>();
        if (!stageData.currentItem().isBlank()) {
            items.add(new MiniCVisualItem(stageData.currentItem(), true));
        }
        source.stream()
                .limit(24)
                .map(item -> new MiniCVisualItem(item, false))
                .forEach(items::add);
        if (items.isEmpty()) {
            items.add(new MiniCVisualItem(stageData.stage() + " 暂无输出", true));
        }
        return List.copyOf(items);
    }

    private List<String> sourceFor(String stage, UiStageDataDto stageData, UiGlobalDataDto globalData) {
        if (globalData == null) {
            return stageData.accumulatedOutput();
        }
        return switch (stage) {
            case "lexer" -> globalData.tokenSummary();
            case "parser" -> globalData.astSummary();
            case "semantic" -> globalData.semanticSummary();
            case "ir" -> globalData.irSummary();
            case "codegen" -> globalData.assemblySummary();
            default -> stageData.accumulatedOutput();
        };
    }
}

File: uiweb/src/visual/MiniCVisualModelFactory.ts
import type { JavaMirrorFile } from "../translation/javaMirror";
import type { UiStageVisualDto } from "../translation/uiapi";
import { MiniCVisualItem } from "./MiniCVisualItem";

export const miniCVisualModelFactoryMirror = {
  javaPath: "src/main/java/minic/uilocal/visual/MiniCVisualModelFactory.java",
  webPath: "uiweb/src/visual/MiniCVisualModelFactory.ts",
  packageName: "minic.uilocal.visual",
  exportName: "MiniCVisualModelFactory",
  kind: "class",
  imports: [
    "java.util.ArrayList",
    "java.util.List",
    "minic.uiapi.UiGlobalDataDto",
    "minic.uiapi.UiStageDataDto"
  ],
  fields: [],
  methods: [
    {
      "name": "create",
      "signature": "create(UiStageDataDto stageData,UiGlobalDataDto globalData)"
    },
    {
      "name": "sourceFor",
      "signature": "sourceFor(String stage,UiStageDataDto stageData,UiGlobalDataDto globalData)"
    }
  ],
} as const satisfies JavaMirrorFile;

export interface MiniCStageDataLike {
  readonly stage: string;
  readonly currentItem: string;
  readonly accumulatedOutput: readonly string[];
}

export interface MiniCGlobalDataLike {
  readonly preprocessSummary?: readonly string[];
  readonly tokenSummary?: readonly string[];
  readonly astSummary?: readonly string[];
  readonly semanticSummary?: readonly string[];
  readonly irSummary?: readonly string[];
  readonly assemblySummary?: readonly string[];
}

export class MiniCVisualModelFactory {
  static readonly mirror = miniCVisualModelFactoryMirror;

  create(stageData: MiniCStageDataLike | null | undefined, globalData?: MiniCGlobalDataLike | null): readonly MiniCVisualItem[] {
    if (!stageData) {
      return [new MiniCVisualItem("等待开始观测会话", true)];
    }
    const source = this.sourceFor(stageData.stage, stageData, globalData ?? null);
    const items: MiniCVisualItem[] = [];
    if (stageData.currentItem.trim().length > 0) {
      items.push(new MiniCVisualItem(stageData.currentItem, true));
    }
    items.push(...source.slice(0, 24).map((item) => new MiniCVisualItem(item, false)));
    if (items.length === 0) {
      items.push(new MiniCVisualItem(`${stageData.stage} 暂无输出`, true));
    }
    return items;
  }

  createFromVisual(visual: UiStageVisualDto | null | undefined): readonly MiniCVisualItem[] {
    if (!visual) {
      return [new MiniCVisualItem("暂无可视化数据", true)];
    }
    const items = [
      ...visual.genericItems,
      ...visual.lexerTokens.map((token) => `${token.kind} ${token.text}`.trim()),
      ...visual.irLines.map((line) => line.text),
      ...visual.assemblyLines.map((line) => line.text),
    ];
    return items.length > 0
      ? items.slice(0, 24).map((item, index) => new MiniCVisualItem(item, index === 0))
      : [new MiniCVisualItem(`${visual.stage} 暂无输出`, true)];
  }

  private sourceFor(stage: string, stageData: MiniCStageDataLike, globalData: MiniCGlobalDataLike | null): readonly string[] {
    if (!globalData) {
      return stageData.accumulatedOutput;
    }
    switch (stage) {
      case "preprocess":
        return globalData.preprocessSummary ?? stageData.accumulatedOutput;
      case "lexer":
        return globalData.tokenSummary ?? stageData.accumulatedOutput;
      case "parser":
        return globalData.astSummary ?? stageData.accumulatedOutput;
      case "semantic":
        return globalData.semanticSummary ?? stageData.accumulatedOutput;
      case "ir":
        return globalData.irSummary ?? stageData.accumulatedOutput;
      case "codegen":
        return globalData.assemblySummary ?? stageData.accumulatedOutput;
      default:
        return stageData.accumulatedOutput;
    }
  }
}

export default MiniCVisualModelFactory;
